<a href="https://colab.research.google.com/github/rocioivonneperezgonzalez-ship-it/Challenge_TlecomX_Alura_LATAM/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Telecom X - Análisis de Evasión de Clientes**


Has sido contratado como asistente de análisis de datos en Telecom X y formarás parte del proyecto "Churn de Clientes". La empresa enfrenta una alta tasa de cancelaciones y necesita comprender los factores que llevan a la pérdida de clientes.
Tu desafío será recopilar, procesar y analizar los datos, utilizando Python y sus principales bibliotecas para extraer información valiosa. A partir de tu análisis, el equipo de Data Science podrá avanzar en modelos predictivos y desarrollar estrategias para reducir la evasión.
¿Qué vas a practicar?
✅ Importar y manipular datos desde una API de manera eficiente.
✅ Aplicar los conceptos de ETL (Extracción, Transformación y Carga) en la preparación de los datos.
✅ Crear visualizaciones estratégicas para identificar patrones y tendencias.
✅ Realizar un Análisis Exploratorio de Datos (EDA) y generar un informe con insights relevantes.

##📌 Extracción

Para iniciar tu análisis, necesitarás importar los datos de la API de Telecom X. Estos datos están disponibles en formato JSON y contienen información esencial sobre los clientes, incluyendo datos demográficos, tipo de servicio contratado y estado de evasión.




***¿Qué debes hacer?***

✅ Cargar los datos directamente desde la API utilizando Python.

✅ Convertir los datos a un DataFrame de Pandas para facilitar su manipulación.

In [1]:
import pandas as pd
import requests
import json

In [2]:
url = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/refs/heads/main/TelecomX_Data.json'
response = requests.get(url)
data = response.json()
df = pd.DataFrame(data)
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [3]:
df = pd.json_normalize(data)  # df plano (normalizado)
df.head()

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.OnlineBackup  \
0                      DSL  ...                   Yes   
1                      DSL  ...                    No   
2              Fiber optic  ...                    No   
3              Fiber optic  ...                   Yes   
4              Fiber optic  ...                    No   

  internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
0          Mailed check                    65.6                  593.3  
1          Mailed check                    59.9                  542.4  
2      Electronic check                    73.9                 280.85  
3      Electronic check                    98.0                1237.85  
4          Mailed check                    83.9                  267.4  

[5 rows x 21 columns]

🔍 **Análisis de la Estructura de los Datos**
Los datos están organizados en un formato jerárquico (diccionario dentro de diccionario), conteniendo la siguiente información:

1. **Identificación del Cliente** - `'customerID'`: Identificador único del cliente.
* `'Churn'`: Indica si el cliente canceló el servicio (`Yes` o `No`).


2. **Información del Cliente** (`'customer'`)
* `'gender'`: Género del cliente (`Male` o `Female`).
* `'SeniorCitizen'`: Indica si el cliente es una persona mayor (0 = No, 1 = Sí).
* `'Partner'`: Si el cliente tiene pareja.
* `'Dependents'`: Si el cliente tiene dependientes.
* `'tenure'`: Tiempo de permanencia como cliente (en meses).


3. **Servicios de Telefonía** (`'phone'`)
* `'PhoneService'`: Indica si el cliente posee servicio de teléfono (`Yes` o `No`).
* `'MultipleLines'`: Si posee múltiples líneas telefónicas.


4. **Servicios de Internet** (`'internet'`)
* Tipo de servicio de internet contratado (`DSL`, `Fiber optic`, `No`).
* Servicios adicionales (`OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`).


5. **Información de la Cuenta** (`'account'`)
* `'Contract'`: Tipo de contrato (`Month-to-month`, `One year`, `Two year`).
* `'PaperlessBilling'`: Si el cliente recibe facturas electrónicas.
* `'PaymentMethod'`: Método de pago (`Bank transfer`, `Credit card`, `Electronic check`, `Mailed check`).
* `'Charges'`:
* `'Monthly'`: Valor mensual cobrado.
* `'Total'`: Valor total pagado por el cliente.

##🔧 Transformación

### Conociendo el conjunto de datos

Ahora que has extraído los datos, es fundamental comprender la estructura del dataset y el significado de sus columnas. Esta etapa te ayudará a identificar qué variables son más relevantes para el análisis de evasión de clientes.

📌 Para facilitar este proceso, hemos creado un diccionario de datos con la descripción de cada columna. Aunque no es obligatorio utilizarlo, puede ayudarte a comprender mejor la información disponible.

🔗 Enlace al diccionario y a la API

¿Qué debes hacer?

✅ Explorar las columnas del dataset y verificar sus tipos de datos.

✅ Consultar el diccionario para comprender mejor el significado de las variables.

✅ Identificar las columnas más relevantes para el análisis de evasión

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [5]:
df.shape
df.columns
df.dtypes
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

### Comprobación de incohernecias en los datos

En este paso, verifica si hay problemas en los datos que puedan afectar el análisis.

Presta atención a valores ausentes, duplicados, errores de formato e inconsistencias en las categorías.

Este proceso es esencial para asegurarte de que los datos estén listos para las siguientes etapas.

VERIFICACIÓN DE VALORES UNICOS

In [6]:
for col in df.columns:
    print(f"{col}: {df[col].unique()}")
    if df[col].nunique()<50:
      print (df[col].unique())
      print('-' * 50)

customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Churn: ['No' 'Yes' '']
['No' 'Yes' '']
--------------------------------------------------
customer.gender: ['Female' 'Male']
['Female' 'Male']
--------------------------------------------------
customer.SeniorCitizen: [0 1]
[0 1]
--------------------------------------------------
customer.Partner: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
customer.Dependents: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
customer.tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
phone.PhoneService: ['Yes' 'No']
['Yes' 'No']
--------------------------------------------------
phone.MultipleLines: ['No' 'Yes' 'No phone service']
['No' 'Yes' 'No phone service']
------

VALORES DUPLICADOS

In [7]:
print(f"Número de filas duplicadas: {df.duplicated().sum()}")

Número de filas duplicadas: 0


*El dataset no presenta registros duplicados, ni filas idénticas ni identificadores customerID repetidos.*

VALORES NULOS

In [8]:
print(f"Número de valores nulos\n: {df.isnull().sum()}")

Número de valores nulos
: customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64


*Lo que nos indica que no existe ningun valor nulo*

VALORES VACIOS O EN BLANCO

In [9]:
df.apply(lambda x: x.astype(str).str.strip()=='').sum()

customerID                     0
Churn                        224
customer.gender                0
customer.SeniorCitizen         0
customer.Partner               0
customer.Dependents            0
customer.tenure                0
phone.PhoneService             0
phone.MultipleLines            0
internet.InternetService       0
internet.OnlineSecurity        0
internet.OnlineBackup          0
internet.DeviceProtection      0
internet.TechSupport           0
internet.StreamingTV           0
internet.StreamingMovies       0
account.Contract               0
account.PaperlessBilling       0
account.PaymentMethod          0
account.Charges.Monthly        0
account.Charges.Total         11
dtype: int64

 *Se tiene 224 valores vacíos/solo-espacios en Churn.*

### Manejo de inconsistencias

1) Corregir formato de Charges.Total (a Float) ✅

Para un mejor manejo de os datos, ya que se pudo observa que se encuentra como object

In [10]:
df['account.Charges.Total'] = pd.to_numeric(df['account.Charges.Total'], errors='coerce')

2. Para un mejor manejo de el DataFrame se procede a aliminar los 224 valores vacio que se encontraron en `Churn`

In [11]:
df = df[df['Churn'].str.strip() !='']
print(f"Número de filas despues de eliminar las avcia en churn:", len(df))

Número de filas despues de eliminar las avcia en churn: 7043


*Una vez eliminadas los 224 valores vacios nos quedamos con 7043 valores*

3. Prodedemos a verificar cuantos valores vacios o en blancos quedaron en el DataFrame despues de eliminar los 224 valores en Churn

In [12]:
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

In [13]:
print("Numero de Nulos \n", df.isnull().sum())

Numero de Nulos 
 customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64


### Columna de cuentas diarias (Opcional)

Ahora que los datos están limpios, es momento de crear la columna "Cuentas_Diarias". Utiliza la facturación mensual para calcular el valor diario, proporcionando una visión más detallada del comportamiento de los clientes a lo largo del tiempo.
📌 Esta columna te ayudará a profundizar en el análisis y a obtener información valiosa para las siguientes etapas.

In [14]:
df["Cuentas_Diarias"] = df["account.Charges.Total"] / 30
df.head()


customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.DeviceProtection  \
0                      DSL  ...                        No   
1                      DSL  ...                        No   
2              Fiber optic  ...                       Yes   
3              Fiber optic  ...                       Yes   
4              Fiber optic  ...                        No   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   
3                   No                  Yes                      Yes   
4                  Yes                  Yes                       No   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         One year                      Yes          Mailed check   
1   Month-to-month                       No          Mailed check   
2   Month-to-month                      Yes      Electronic check   
3   Month-to-month                      Yes      Electronic check   
4   Month-to-month                      Yes          Mailed check   

  account.Charges.Monthly account.Charges.Total  Cuentas_Diarias  
0                    65.6                593.30        19.776667  
1                    59.9                542.40        18.080000  
2                    73.9                280.85         9.361667  
3                    98.0               1237.85        41.261667  
4                    83.9                267.40         8.913333  

[5 rows x 22 columns]

In [15]:
df["Cuentas_Diarias"].describe()


count    7032.000000
mean       76.110015
std        75.559045
min         0.626667
25%        13.381667
50%        46.582500
75%       126.491250
max       289.493333
Name: Cuentas_Diarias, dtype: float64

### Estandarización y tranformación de datos (Opcional)

Para un mejor manejo de columnas reasignaremos nombres a éstas

In [16]:
col = {                                                        # Diccionario con los nuevos nombres de las columnas.
    'customerID': 'id',
    'customer.gender': 'genero',
    'customer.SeniorCitizen': 'mayor_de_60',
    'customer.Partner': 'posee_pareja',
    'customer.Dependents': 'posee_dependientes',
    'customer.tenure': 'tiempo_contrato',
    'phone.PhoneService': 'servicio_telefono',
    'phone.MultipleLines': 'multiples_lineas',
    'internet.InternetService': 'tipo_internet',
    'internet.OnlineSecurity': 'seguridad_online',
    'internet.OnlineBackup': 'backup_online',
    'internet.DeviceProtection': 'proteccion_dispositivo',
    'internet.TechSupport': 'soporte_tecnico',
    'internet.StreamingTV': 'streaming_tv',
    'internet.StreamingMovies': 'streaming_peliculas',
    'account.Contract': 'tipo_contrato',
    'account.PaperlessBilling': 'factura_digital',
    'account.PaymentMethod': 'metodo_pago',
    'account.Charges.Monthly': 'valor_mensal',
    'account.Charges.Total': 'total_cobrado'
}
df = df.rename(columns= col)
df

id Churn  genero  mayor_de_60 posee_pareja posee_dependientes  \
0     0002-ORFBO    No  Female            0          Yes                Yes   
1     0003-MKNFE    No    Male            0           No                 No   
2     0004-TLHLJ   Yes    Male            0           No                 No   
3     0011-IGKFF   Yes    Male            1          Yes                 No   
4     0013-EXCHZ   Yes  Female            1          Yes                 No   
...          ...   ...     ...          ...          ...                ...   
7262  9987-LUTYD    No  Female            0           No                 No   
7263  9992-RRAMN   Yes    Male            0          Yes                 No   
7264  9992-UJOEL    No    Male            0           No                 No   
7265  9993-LHIEB    No    Male            0          Yes                Yes   
7266  9995-HOTOH    No    Male            0          Yes                Yes   

      tiempo_contrato servicio_telefono  multiples_lineas tipo_internet  ...  \
0                   9               Yes                No           DSL  ...   
1                   9               Yes               Yes           DSL  ...   
2                   4               Yes                No   Fiber optic  ...   
3                  13               Yes                No   Fiber optic  ...   
4                   3               Yes                No   Fiber optic  ...   
...               ...               ...               ...           ...  ...   
7262               13               Yes                No           DSL  ...   
7263               22               Yes               Yes   Fiber optic  ...   
7264                2               Yes                No           DSL  ...   
7265               67               Yes                No           DSL  ...   
7266               63                No  No phone service           DSL  ...   

     proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                        No             Yes          Yes                  No   
1                        No              No           No                 Yes   
2                       Yes              No           No                  No   
3                       Yes              No          Yes                 Yes   
4                        No             Yes          Yes                  No   
...                     ...             ...          ...                 ...   
7262                     No             Yes           No                  No   
7263                     No              No           No                 Yes   
7264                     No              No           No                  No   
7265                    Yes             Yes           No                 Yes   
7266                    Yes              No          Yes                 Yes   

       tipo_contrato factura_digital       metodo_pago valor_mensal  \
0           One year             Yes      Mailed check        65.60   
1     Month-to-month              No      Mailed check        59.90   
2     Month-to-month             Yes  Electronic check        73.90   
3     Month-to-month             Yes  Electronic check        98.00   
4     Month-to-month             Yes      Mailed check        83.90   
...              ...             ...               ...          ...   
7262        One year              No      Mailed check        55.15   
7263  Month-to-month             Yes  Electronic check        85.10   
7264  Month-to-month             Yes      Mailed check        50.30   
7265        Two year              No      Mailed check        67.85   
7266        Two year              No  Electronic check        59.00   

     total_cobrado  Cuentas_Diarias  
0           593.30        19.776667  
1           542.40        18.080000  
2           280.85         9.361667  
3          1237.85        41.261667  
4           267.40         8.913333  
...            ...              ...  
7262        742.90      

Tambien remplazaremos pasaremos la información de ingles a español

In [17]:
df['Churn'] = df['Churn'].replace({'No': '0', 'Yes': '1'})
df['genero'] = df['genero'].replace({'Female': 'Femenino', 'Male': 'Masculino'})
df['posee_pareja'] = df['posee_pareja'].replace({'Yes': '1', 'No': '0'})
df['posee_dependientes'] = df['posee_dependientes'].replace({'Yes': '1', 'No': '0'})
df['servicio_telefono'] = df['servicio_telefono'].replace({'Yes': '1', 'No': '0'})
df['multiples_lineas'] = df['multiples_lineas'].replace({'No': '0', 'Yes': '1', 'No phone service': 'Sin servicio de teléfono'})
df['tipo_internet'] = df['tipo_internet'].replace({'No': '0'})
df['tipo_contrato'] = df['tipo_contrato'].replace({'One year': 'Anual', 'Month-to-month': 'Mensual', 'Two year': 'Bienal'})
df.head()

id Churn     genero  mayor_de_60 posee_pareja posee_dependientes  \
0  0002-ORFBO     0   Femenino            0            1                  1   
1  0003-MKNFE     0  Masculino            0            0                  0   
2  0004-TLHLJ     1  Masculino            0            0                  0   
3  0011-IGKFF     1  Masculino            1            1                  0   
4  0013-EXCHZ     1   Femenino            1            1                  0   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                 1                0           DSL  ...   
1                9                 1                1           DSL  ...   
2                4                 1                0   Fiber optic  ...   
3               13                 1                0   Fiber optic  ...   
4                3                 1                0   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                     No             Yes          Yes                  No   
1                     No              No           No                 Yes   
2                    Yes              No           No                  No   
3                    Yes              No          Yes                 Yes   
4                     No             Yes          Yes                  No   

  tipo_contrato factura_digital       metodo_pago valor_mensal total_cobrado  \
0         Anual             Yes      Mailed check         65.6        593.30   
1       Mensual              No      Mailed check         59.9        542.40   
2       Mensual             Yes  Electronic check         73.9        280.85   
3       Mensual             Yes  Electronic check         98.0       1237.85   
4       Mensual             Yes      Mailed check         83.9        267.40   

   Cuentas_Diarias  
0        19.776667  
1        18.080000  
2         9.361667  
3        41.261667  
4         8.913333  

[5 rows x 22 columns]

In [18]:
columnas_a_traducir = ['seguridad_online', 'backup_online', 'proteccion_dispositivo', 'soporte_tecnico', 'streaming_tv', 'streaming_peliculas']
mapeo = {'No': '0', 'Yes': '1', 'No internet service': 'Sin servicio de internet'}

for col in columnas_a_traducir:
    df[col] = df[col].replace(mapeo)

df.head()

id Churn     genero  mayor_de_60 posee_pareja posee_dependientes  \
0  0002-ORFBO     0   Femenino            0            1                  1   
1  0003-MKNFE     0  Masculino            0            0                  0   
2  0004-TLHLJ     1  Masculino            0            0                  0   
3  0011-IGKFF     1  Masculino            1            1                  0   
4  0013-EXCHZ     1   Femenino            1            1                  0   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                 1                0           DSL  ...   
1                9                 1                1           DSL  ...   
2                4                 1                0   Fiber optic  ...   
3               13                 1                0   Fiber optic  ...   
4                3                 1                0   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                      0               1            1                   0   
1                      0               0            0                   1   
2                      1               0            0                   0   
3                      1               0            1                   1   
4                      0               1            1                   0   

  tipo_contrato factura_digital       metodo_pago valor_mensal total_cobrado  \
0         Anual             Yes      Mailed check         65.6        593.30   
1       Mensual              No      Mailed check         59.9        542.40   
2       Mensual             Yes  Electronic check         73.9        280.85   
3       Mensual             Yes  Electronic check         98.0       1237.85   
4       Mensual             Yes      Mailed check         83.9        267.40   

   Cuentas_Diarias  
0        19.776667  
1        18.080000  
2         9.361667  
3        41.261667  
4         8.913333  

[5 rows x 22 columns]

In [19]:
df['metodo_pago'] = df['metodo_pago'].replace({
    'Mailed check': 'Cheque enviado por correo',
    'Electronic check': 'Cheque electrónico',
    'Credit card (automatic)': 'Tarjeta de crédito (automático)',
    'Bank transfer (automatic)': 'Transferencia bancaria (automática)'
})
df.head()

id Churn     genero  mayor_de_60 posee_pareja posee_dependientes  \
0  0002-ORFBO     0   Femenino            0            1                  1   
1  0003-MKNFE     0  Masculino            0            0                  0   
2  0004-TLHLJ     1  Masculino            0            0                  0   
3  0011-IGKFF     1  Masculino            1            1                  0   
4  0013-EXCHZ     1   Femenino            1            1                  0   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                 1                0           DSL  ...   
1                9                 1                1           DSL  ...   
2                4                 1                0   Fiber optic  ...   
3               13                 1                0   Fiber optic  ...   
4                3                 1                0   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                      0               1            1                   0   
1                      0               0            0                   1   
2                      1               0            0                   0   
3                      1               0            1                   1   
4                      0               1            1                   0   

  tipo_contrato factura_digital                metodo_pago valor_mensal  \
0         Anual             Yes  Cheque enviado por correo         65.6   
1       Mensual              No  Cheque enviado por correo         59.9   
2       Mensual             Yes         Cheque electrónico         73.9   
3       Mensual             Yes         Cheque electrónico         98.0   
4       Mensual             Yes  Cheque enviado por correo         83.9   

  total_cobrado  Cuentas_Diarias  
0        593.30        19.776667  
1        542.40        18.080000  
2        280.85         9.361667  
3       1237.85        41.261667  
4        267.40         8.913333  

[5 rows x 22 columns]

In [20]:
df

id Churn     genero  mayor_de_60 posee_pareja  \
0     0002-ORFBO     0   Femenino            0            1   
1     0003-MKNFE     0  Masculino            0            0   
2     0004-TLHLJ     1  Masculino            0            0   
3     0011-IGKFF     1  Masculino            1            1   
4     0013-EXCHZ     1   Femenino            1            1   
...          ...   ...        ...          ...          ...   
7262  9987-LUTYD     0   Femenino            0            0   
7263  9992-RRAMN     1  Masculino            0            1   
7264  9992-UJOEL     0  Masculino            0            0   
7265  9993-LHIEB     0  Masculino            0            1   
7266  9995-HOTOH     0  Masculino            0            1   

     posee_dependientes  tiempo_contrato servicio_telefono  \
0                     1                9                 1   
1                     0                9                 1   
2                     0                4                 1   
3                     0               13                 1   
4                     0                3                 1   
...                 ...              ...               ...   
7262                  0               13                 1   
7263                  0               22                 1   
7264                  0                2                 1   
7265                  1               67                 1   
7266                  1               63                 0   

              multiples_lineas tipo_internet  ... proteccion_dispositivo  \
0                            0           DSL  ...                      0   
1                            1           DSL  ...                      0   
2                            0   Fiber optic  ...                      1   
3                            0   Fiber optic  ...                      1   
4                            0   Fiber optic  ...                      0   
...                        ...           ...  ...                    ...   
7262                         0           DSL  ...                      0   
7263                         1   Fiber optic  ...                      0   
7264                         0           DSL  ...                      0   
7265                         0           DSL  ...                      1   
7266  Sin servicio de teléfono           DSL  ...                      1   

     soporte_tecnico streaming_tv streaming_peliculas tipo_contrato  \
0                  1            1                   0         Anual   
1                  0            0                   1       Mensual   
2                  0            0                   0       Mensual   
3                  0            1                   1       Mensual   
4                  1            1                   0       Mensual   
...              ...          ...                 ...           ...   
7262               1            0                   0         Anual   
7263               0            0                   1       Mensual   
7264               0            0                   0       Mensual   
7265               1            0                   1        Bienal   
7266               0            1                   1        Bienal   

     factura_digital                metodo_pago valor_mensal total_cobrado  \
0                Yes  Cheque enviado por correo        65.60        593.30   
1                 No  Cheque enviado por correo        59.90        542.40   
2                Yes         Cheque electrónico        73.90        280.85   
3                Yes         Cheque electrónico        98.00       1237.85   
4                Yes  Cheque enviado por correo        83.90        267.40   
...              ...                        ...          ...           ...   
7262              No  Cheque enviado por correo        55.15        742.90   
7263             Yes         Cheque electrónico        85.10       1873.70   
7264             Yes  Cheque enviado por cor

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      7043 non-null   object 
 1   Churn                   7043 non-null   object 
 2   genero                  7043 non-null   object 
 3   mayor_de_60             7043 non-null   int64  
 4   posee_pareja            7043 non-null   object 
 5   posee_dependientes      7043 non-null   object 
 6   tiempo_contrato         7043 non-null   int64  
 7   servicio_telefono       7043 non-null   object 
 8   multiples_lineas        7043 non-null   object 
 9   tipo_internet           7043 non-null   object 
 10  seguridad_online        7043 non-null   object 
 11  backup_online           7043 non-null   object 
 12  proteccion_dispositivo  7043 non-null   object 
 13  soporte_tecnico         7043 non-null   object 
 14  streaming_tv            7043 non-null   objec

##📊 Carga y análisis

1. ANALISIS DECRIPTIVO

Para comenzar, realiza un análisis descriptivo de los datos, calculando métricas como media, mediana, desviación estándar y otras medidas que ayuden a comprender mejor la distribución y el comportamiento de los clientes. 📌 Consejos: 🔗 Documentación de DataFrame.describe()

In [22]:
# Análisis descriptivo general (incluye numéricas)
df.describe()

mayor_de_60  tiempo_contrato  valor_mensal  total_cobrado  \
count  7043.000000      7043.000000   7043.000000    7032.000000   
mean      0.162147        32.371149     64.761692    2283.300441   
std       0.368612        24.559481     30.090047    2266.771362   
min       0.000000         0.000000     18.250000      18.800000   
25%       0.000000         9.000000     35.500000     401.450000   
50%       0.000000        29.000000     70.350000    1397.475000   
75%       0.000000        55.000000     89.850000    3794.737500   
max       1.000000        72.000000    118.750000    8684.800000   

       Cuentas_Diarias  
count      7032.000000  
mean         76.110015  
std          75.559045  
min           0.626667  
25%          13.381667  
50%          46.582500  
75%         126.491250  
max         289.493333

2. Distribución de evasión (Churn)

En este paso, el objetivo es comprender cómo está distribuida la variable "churn" (evasión) entre los clientes. Utiliza gráficos para visualizar la proporción de clientes que permanecieron y los que se dieron de baja.

3. Recuento de evasión por variables categoricas

Ahora, exploraremos cómo se distribuye la evasión según variables categóricas, como género, tipo de contrato, método de pago, entre otras. Este análisis puede revelar patrones interesantes, por ejemplo, si los clientes de ciertos perfiles tienen una mayor tendencia a cancelar el servicio, lo que ayudará a orientar acciones estratégicas.

### Conteo de evasión por variables numericas

En este paso, explora cómo las variables numéricas, como "total gastado" o "tiempo de contrato", se distribuyen entre los clientes que cancelaron (evasión) y los que no cancelaron. Este análisis ayuda a entender si ciertos valores numéricos están más asociados con la evasión, proporcionando insights sobre los factores que influyen en el comportamiento de los clientes.

##📄Informe final